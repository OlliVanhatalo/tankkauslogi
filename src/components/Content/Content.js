//Sisältökomponentti. Luotu muotoilun helpottamiseksi. Sisään
//kirjoitettuihin komponentteihin tuodaan kaikki niiden
//ominaisuudet propsilla
//-----------------------------------------------------------

import React from 'react';

import './Content.css';

function Content(props) {
    return (
      <div className="content">
        { props.children }
  
      </div>
    )
  }

export default Content;