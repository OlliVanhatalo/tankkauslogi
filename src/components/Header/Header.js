//Muotoilukomponentti. Sisältää sivun headerin sisällön
//-----------------------------------------------------------

import React from 'react';

import './Header.css'

function Header(props) {
    return(
      <div className="header">
        <h1>Tankkauslogi</h1>
      </div>
    )
}

export default Header;