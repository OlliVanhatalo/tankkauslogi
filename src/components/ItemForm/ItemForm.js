import React from 'react';
import { withRouter } from 'react-router';
import { v4 as uuidv4 } from 'uuid';

import Button from '../buttons';

import './ItemForm.css'

class ItemForm extends React.Component {

    constructor(props) {
        super(props);
        const data = props.data ? props.data : {
            tyyppi: "95E10",
            summa: 0,
            litraa: 0,
            tankkauspaiva: "",
            ajoneuvo: "",
            litrahinta: ""
         }
        this.state = {
            data: data
        };
        this.handleInputChange = this.handleInputChange.bind(this);
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleCancel = this.handleCancel.bind(this);
        this.handleDeleteItem = this.handleDeleteItem.bind(this);
    }

    handleInputChange(event) {
        const target = event.target;
        const value = target.name === 'isGoing' ? target.checked : target.value;
        const name = target.name;
    
        this.setState({
          data: {
              ...this.state.data,
              [name]: value
          }
        });
    }

    handleCancel(event) {
        event.preventDefault();
        this.props.history.goBack();
    }

    handleSubmit(event) {
        event.preventDefault();
        let data = Object.assign({}, this.state.data);
        data.summa = parseFloat(data.summa);
        data.id = data.id ? data.id : uuidv4();
        this.props.onFormSubmit(data);
        this.props.history.push("/");
    }

    handleDeleteItem(event) {
        event.preventDefault();
        this.props.onDeleteItem(this.state.data.id);
        this.props.history.push("/");
    }


    render() {
        return(
            <form onSubmit={this.handleSubmit}>

            <div className="itemform">

                <div className="itemform__row">
                <div>
                    <label htmlFor="tyyppi">Polttoainetyyppi</label>
                    <select name="tyyppi" value={this.state.data.tyyppi} onChange={this.handleInputChange}>

                        {this.props.selectList.map(item => <option value={item} key={item}>{item}</option>)}

                    </select>
                </div>
                </div>

                <div className="itemform__row">
                <div>
                    <label htmlFor="summa">Summa</label>
                    <input type="number" name="summa" step="0.01" value={this.state.data.summa} onChange={this.handleInputChange}/>
                </div>

                <div>
                    <label htmlFor="litraa">Litraa</label>
                    <input type="number" name="litraa" value={this.state.data.litraa} onChange={this.handleInputChange}/>
                </div>
                </div>

                <div className="itemform__row">
                <div>
                    <label htmlFor="tankkauspaiva">Tankkauspäivä</label>
                    <input type="date" name="tankkauspaiva" size="10" value={this.state.data.tankkauspaiva} onChange={this.handleInputChange}/>
                </div>
                </div>

                <div className="itemform__row">
                <div>
                <label htmlFor="ajoneuvo">Ajoneuvo</label>
                    <select name="ajoneuvo" value={this.state.data.ajoneuvo} onChange={this.handleInputChange}>

                        {this.props.vehicle.map(item => <option value={item} key={item}>{item}</option>)}

                    </select>
                </div>
                </div>

                <div className="itemform__row">
                <div>
                    <Button onClick={this.handleCancel}>PERUUTA</Button>
                </div>
                <div>
                    <Button type="submit" primary>{this.state.data.id ? "TALLENNA" : "LISÄÄ"}</Button>
                </div>
                </div>

                { this.props.onDeleteItem ? 
                    <div className="itemform__row">
                        <div>
                            <Button onClick={this.handleDeleteItem}>POISTA</Button> 
                        </div>
                        <div></div>
                    </div> : "" }

            </div>

            </form>
        );
    }
}

export default withRouter(ItemForm);