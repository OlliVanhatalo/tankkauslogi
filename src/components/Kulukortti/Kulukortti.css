.kulukortti {
    border-bottom: 1px solid rgba(224, 130, 130, 0.575);
    display: flex;
    background-color: rgba(255, 255, 255, 0.35);
    border-radius: 1rem;
  }
  
  .kulukortti__rivi {
    display: flex;
    justify-content: space-between;
  }
  
  .kulukortti__tyyppi, .kulukortti__summa {
    font-weight: bold;
  }
  
  .kulukortti__tankkauspaiva, .kulukortti__litraa, .kulukortti__ajoneuvo, .kulukortti__litrahinta {
    font-size: 85%;
    color: #666;
  
  }

  .kulukortti > div {
    flex: 1 1 auto;
    padding: 1em;
  }

  .kulukortti__linkki {
    max-width: 1em;
    display: none;
    justify-content: center;
    align-items: center;
    background-color: rgba(0, 0, 0, 0.1);
  }

  .kulukortti:hover .kulukortti__linkki {
    display: flex;
  }

  .kulukortti__linkki a {
    color: #000;
  }