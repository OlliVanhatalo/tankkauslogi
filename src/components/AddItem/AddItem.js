//Uuden tankkaustapahtuman lis채채minen. Komponenttiin on tuotu 
//lomakekomponentti
//-----------------------------------------------------------

import React from 'react';

import Content from '../Content/Content';
import ItemForm from '../ItemForm/ItemForm';


import './AddItem.css';

function AddItem(props) {
    return (
      <Content>

        <div className="additem">

        <h2>Uuden tankkauksen lis채채minen</h2>

        <ItemForm onFormSubmit={props.onFormSubmit} selectList={props.selectList} vehicle={props.vehicle} />

        </div>
      </Content>
    )
  }

export default AddItem;