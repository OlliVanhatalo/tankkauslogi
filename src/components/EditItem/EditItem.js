//Tankkaustapahtuman muokkaaminen. Vastaava kuin AddItem. 
//Lisäominaisuutena lomaketiedon tuonti jo tehdystä tapahtumasta, 
//sekä lomakkeen poistamisen mahdollistavan toiminnallisuuden tuonti
//-----------------------------------------------------------

import React from 'react';

import Content from '../Content/Content';
import ItemForm from '../ItemForm/ItemForm';


import './EditItem.css';

function EditItem(props) {

    const index = props.data.findIndex(item => item.id === props.match.params.id);

    let itemData = props.data[index];

    return (
      <Content>

        <div className="edititem">

        <h2>Tankkauksen muokkaaminen</h2>

        <ItemForm onFormSubmit={props.onFormSubmit} 
                  selectList={props.selectList} 
                  data={itemData} 
                  onDeleteItem={props.onDeleteItem} 
                  vehicle={props.vehicle}/>

        </div>
      </Content>
    )
  }

export default EditItem;