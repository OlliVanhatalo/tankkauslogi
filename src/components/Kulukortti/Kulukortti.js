//Tankkauskortin tiedot tulostava komponentti. Data on jaoteltu
//riveihin div-elementeillä muotoilun helpottamiseksi.
//-----------------------------------------------------------

import React from 'react';
import moment from 'moment';

import ArrowRight from '@material-ui/icons/ArrowRight';
import { Link } from 'react-router-dom';

import './Kulukortti.css'

//Funktio tulostaa kulukortin sisällön. Lisäksi määritellään 
//tankkauspäivän muotoilu moment.js funktion avulla, sekä määritellään 
//muuttuja "litrahinta", jolle määritetään arvo if-lauseella
//hyödyntäen tankkauskortista löytyvää summaa ja litraa.
function Kulukortti(props){

    let tankkauspaiva = moment(props.data.tankkauspaiva);
    let litrahinta = "";

    if (props.data.summa && props.data.litraa) {
      litrahinta = props.data.summa / props.data.litraa; 
    } else { litrahinta = 0; }

    //Kulukortin rakenne ja sisältö riveittäin
    return(
      <div className="kulukortti">
        <div className="kulukortti__ryhma">
        <div className="kulukortti__rivi">
          <div className="kulukortti__tyyppi">{props.data.tyyppi}</div>
          <div className="kulukortti__summa">{props.data.summa.toFixed(2)} €</div>
        </div>
        <div className="kulukortti__rivi">
          <div className="kulukortti__tankkauspaiva">{tankkauspaiva.format("D.M.Y")}</div>
          <div className="kulukortti__litraa">{props.data.litraa} l</div>
        </div>
        <div className="kulukortti__rivi">
          <div className="kulukortti__ajoneuvo">{props.data.ajoneuvo}</div> 
          <div className="kulukortti__litrahinta">{litrahinta.toFixed(3)}€ / l</div>  
        </div>
        </div>
        <div className="kulukortti__linkki">
          <Link to={"/edit/" + props.data.id}><ArrowRight /></Link>
        </div>
      </div>
    );
  }  

export default Kulukortti;