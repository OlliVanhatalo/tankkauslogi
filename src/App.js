import React, { Component} from 'react';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import './App.css';
import firebase, { auth, provider } from './firebase';

import Header from './components/Header/Header';
import Items from './components/Items/Items';
//import Vehicles from './components/Vehicles/Vehicles';
import Stats from './components/Stats/Stats';
import Settings from './components/Settings/Settings';
import Menu from './components/Menu/Menu';
import AddItem from './components/AddItem/AddItem';
//import AddVehicle from './components/AddVehicle/AddVehicle';
import EditItem from './components/EditItem/EditItem';
//import EditVehicle from './components/EditVehicle/EditVehicle';
import Content from './components/Content/Content';
import Button from './components/buttons';
import vehicledata from './vehicledata';

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      data: [],
      selectList: ["95E10","98E5", "DI", "E85" ],
      vehicle: ['Auto 1','Auto 2'],
      user: null,
      error: null
    }
    this.dbRef = firebase.firestore();
    this.handleFormSubmit = this.handleFormSubmit.bind(this);
    this.handleSelectListForm = this.handleSelectListForm.bind(this);
    this.handleVehicleForm = this.handleVehicleForm.bind(this);
    this.handleDeleteItem = this.handleDeleteItem.bind(this);
    this.login = this.login.bind(this);
    this.logout = this.logout.bind(this);
  }

  componentDidMount() {
    auth.onAuthStateChanged((user) => {
      if (user) {
        this.setState({ 
          user: user 
        });

        //Luodaan tallennettavalle datalle osoitteet tietokantaan
        this.refData = this.dbRef.collection("users").doc(user.uid).collection('data');
        
        //Luodaan tankkausten ja ajoneuvojen datalle tietokokoelmat tietokannassa,
        //käydään läpi siellä olevat tiedot ja tuodaan ne App.js:än this.stateen.
        this.unsubscribe = this.refData.orderBy("tankkauspaiva","desc").onSnapshot((docs) => {
          let data = [];
          docs.forEach((doc) => {
            let docdata =  doc.data();
            data.push(docdata);
            
          });
          this.setState({
            data: data
          });
        });

        //Tämä osuus liityy uusien ajoneuvojen tallentamiseen tietokantaan,
        //mutta en lopultakaan saanut sitä toimimaan
        //------------------------------------------------------------------
        // this.refVehicleData = this.dbRef.collection("users").doc(user.uid).collection('vehicledata');

        // this.unsubscribe = this.refData.onSnapshot((docs) => {
        //   let vehicle = [];
        //   docs.forEach((doc) => {
        //     let docdata =  doc.data();
        //     vehicle.push(vehicle);
        //   });
        //   this.setState({
        //     vehicle: vehicle
        //   });
        // });
        //------------------------------------------------------------------
      } 
    }); 
  }

  handleFormSubmit(newdata) {
    this.refData.doc(newdata.id).set(newdata);

    //Tämä pätkä liittyy myös ajoneuvojen tallentamiseen, joka ei toiminut
    //--------------------------------------------------------------------
    // this.refVehicleData.doc(newdata.id).set(newdata);
    //--------------------------------------------------------------------
  }

  handleSelectListForm(newitem) {
    let selectList = this.state.selectList.slice();
    selectList.push(newitem);
    selectList.sort();
    this.setState({
      selectList: selectList
    }); 
  }

  handleVehicleForm(newitem) {
    let vehicle = this.state.vehicle.slice();
    vehicle.push(newitem);
    vehicle.sort();
    this.setState({
      vehicle: vehicle
    }); 
  }

  handleDeleteItem(id) {
    this.refData.doc(id).delete().then().catch(error => {console.error("Virhe tietoa poistettaessa: ", error)})
    //this.vehicleData.doc(id).delete().then().catch(error => {console.error("Virhe tietoa poistettaessa: ", error)})
  }

  login() {
    auth.signInWithPopup(provider).then((result) => {
      const user = result.user;
      this.setState({
        user: user,
        error: null
      });
    }).catch((error) => {
      const errorMessage = error.message;
      this.setState({
        error: errorMessage
      });
    });
  }

  logout() {
    this.unsubscribe();
    auth.signOut().then(() => {
      this.setState({
        user: null
      });
      this.refData = null
    });  
  }

  render() {

    if (!this.state.user) {
      return (
        <Router>
          <div className="App">
            <Header/>
            <Content>
            <div className="App__welcome">
                <div>Et ole vielä kirjautunut sisälle.</div>
                <div><Button primary onClick={this.login}>Kirjaudu sisälle</Button></div>
                <div>{this.state.error ? <p>{this.state.error}</p> : null}</div>
              </div>
            </Content>
            <Menu />
          </div>
        </Router>
      );
    }

    return (
      <Router>
        <div className="App">
          <Header/>
          <Route path="/" exact render={() => <Items data={this.state.data} />} />
          <Route path="/stats" render={() => <Stats data={this.state.data} />} />
          <Route path="/settings" render={() => <Settings data={this.state.vehicle}  
                                                          onFormSubmit={this.handleVehicleForm}
                                                          onLogout={this.logout}
                                                          user={this.state.user}
                                                          />} />
          <Route path="/add" render={() => <AddItem onFormSubmit={this.handleFormSubmit} 
                                                    selectList={this.state.selectList} 
                                                    vehicle={this.state.vehicle} />} />
          <Route path="/edit/:id" render={(props) => <EditItem data={this.state.data} 
                                                               selectList={this.state.selectList}
                                                               vehicle={this.state.vehicle}
                                                               onFormSubmit={this.handleFormSubmit} 
                                                               onDeleteItem={this.handleDeleteItem}
                                                               {...props} />} />


          {/*Nämä reititykset liittyvät lisättävien ajoneuvojen korttirakenteeseen. Korttirakenteessa 
          on jotain oleellista vialla, enkä ole vielä keksinyt mikä. Koodit on jätetty olemaan
          myöhempää ongelman ratkaisua varten.
          ------------------------------------------------------------------------------------------- 
          <Route path="/vehicles" render={() => <Vehicles data={this.state.vehicle} />} />          
          <Route path="/addvehicle" render={() => <AddVehicle onFormSubmit={this.handleFormSubmit} 
                                                              selectList={this.state.selectList} 
                                                              vehicle={this.state.vehicle} />} />
          <Route path="/editvehicle/:id" render={(props) => <EditVehicle data={this.state.data}
                                                               vehicle={this.state.vehicle}
                                                               onFormSubmit={this.handleFormSubmit} 
                                                               onDeleteItem={this.handleDeleteItem}
                                                               {...props} />} /> 
          ------------------------------------------------------------------------------------------- */}
          <Menu />
        </div>
      </Router>
    );
  }
}

export default App;

