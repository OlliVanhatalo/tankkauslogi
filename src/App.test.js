import React from 'react';
import App from './App';
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer';
import Stats from './components/Stats/Stats';
import Settings from './components/Settings/Settings';
import Items from './components/Items/Items';
import ItemForm from './components/ItemForm/ItemForm';
import AddItem from './components/AddItem/AddItem';
import EditItem from './components/EditItem/EditItem';
import Content from './components/Content/Content';
import Buttons from './components/buttons/index';
import Header from './components/Header/Header';
import Kulukortti from './components/Kulukortti/Kulukortti';
import Menu from './components/Menu/Menu';

test('Renderöi sivulle tekstin Tankkauslogi', () => {
    const { getByText } = render(
    <App />,  
    <Stats/>,
    <Settings/>
    );
    const headerElement = getByText(/Tankkauslogi/i);
    expect(headerElement).toBeInTheDocument();
  });

//TODO!!
// test('Painaa etusivun plus painiketta ja avaa AddItem komponentin', () => {
//     const { getByText } = render(<App />); 
//     const headerElement = getByText(/Tankkauslogi/i);
//     expect(headerElement).toBeInTheDocument();
//   });

  test('Kaikki komponentit renderöityy oikein, verrataan snapshottiin', () => {
    const component = renderer.create(
      <App />,
      <Stats />,
      <Settings />,
      <Items />,
      <ItemForm />,
      <AddItem />,
      <EditItem />,
      <Content />,
      <Buttons />,
      <Header />,
      <Kulukortti />,
      <Menu />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });