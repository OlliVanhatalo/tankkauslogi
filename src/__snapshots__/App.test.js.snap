// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kaikki komponentit renderöityy oikein, verrataan snapshottiin 1`] = `
<div
  className="App"
>
  <div
    className="header"
  >
    <h1>
      Tankkauslogi
    </h1>
  </div>
  <div
    className="content"
  >
    <div
      className="App__welcome"
    >
      <div>
        Et ole vielä kirjautunut sisälle.
      </div>
      <div>
        <button
          className="button  button--primary "
          onClick={[Function]}
          type="button"
        >
          Kirjaudu sisälle
        </button>
      </div>
      <div />
    </div>
  </div>
  <div
    className="menu"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <div
        className="menu__nappi"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          color="#fff"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M4 14h4v-4H4v4zm0 5h4v-4H4v4zM4 9h4V5H4v4zm5 5h12v-4H9v4zm0 5h12v-4H9v4zM9 5v4h12V5H9z"
          />
        </svg>
      </div>
    </a>
    <a
      href="/stats"
      onClick={[Function]}
    >
      <div
        className="menu__nappi"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          color="#fff"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 6l2.29 2.29-4.88 4.88-4-4L2 16.59 3.41 18l6-6 4 4 6.3-6.29L22 12V6z"
          />
        </svg>
      </div>
    </a>
    <a
      href="/settings"
      onClick={[Function]}
    >
      <div
        className="menu__nappi"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root"
          color="#fff"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M15.95 10.78c.03-.25.05-.51.05-.78s-.02-.53-.06-.78l1.69-1.32c.15-.12.19-.34.1-.51l-1.6-2.77c-.1-.18-.31-.24-.49-.18l-1.99.8c-.42-.32-.86-.58-1.35-.78L12 2.34c-.03-.2-.2-.34-.4-.34H8.4c-.2 0-.36.14-.39.34l-.3 2.12c-.49.2-.94.47-1.35.78l-1.99-.8c-.18-.07-.39 0-.49.18l-1.6 2.77c-.1.18-.06.39.1.51l1.69 1.32c-.04.25-.07.52-.07.78s.02.53.06.78L2.37 12.1c-.15.12-.19.34-.1.51l1.6 2.77c.1.18.31.24.49.18l1.99-.8c.42.32.86.58 1.35.78l.3 2.12c.04.2.2.34.4.34h3.2c.2 0 .37-.14.39-.34l.3-2.12c.49-.2.94-.47 1.35-.78l1.99.8c.18.07.39 0 .49-.18l1.6-2.77c.1-.18.06-.39-.1-.51l-1.67-1.32zM10 13c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
            transform="scale(1.2, 1.2)"
          />
        </svg>
      </div>
    </a>
  </div>
</div>
`;
